### DISCUSSION TITLE
Does ntpd-rs have a notion of a system peer?

### DISCUSSION THREAD

**Discussion by paulgear** [pendulum-project/ntpd-rs#1078] [General]
I ask this because from the data I'm collecting from the metrics endpoint it seems to imply some notion of tracking a different host on the basis of the reported stratum.  Here is a snapshot of my metrics since I started monitoring the endpoint: https://snapshots.raintank.io/dashboard/snapshot/1aYZ0MsowSAVRTfdjfzf2kGcWTIL6i5g?orgId=2

Notice how the reported stratum has been jumping between 2 and 3 reasonably frequently over the last couple of days, and how when it initially jumped to stratum 3 for an extended period, this resulted in a large jump in root dispersion, along with some interesting skewing of the peers towards higher offsets, both lagging behind the initial jump to stratum 3.

If there is a system peer, can it be exposed in `ntp-ctl status` and in the metrics endpoint?

**Comment by davidv1992** [comment]
There is not really a notion of system peer, however to comply with the ntpv4 spec ntpd-rs chooses one of it's peers as it's primary source for calculating both the reference id and the stratum. For the time synchronization it by default does a weighted average between the sources that are marked as acceptable.

The spike in dispersion likely has a different reason. Likely for some reason one of the sources became a lot less reliable around that point in time. You can likely see which by looking at the source uncertainties.

**Reply by paulgear** [reply]
Thanks @davidv1992. I'm not seeing any obvious causes.  Here's the same dashboard with peer uncertainties and unanswered polls added: https://snapshots.raintank.io/dashboard/snapshot/wCw7Ei13E24eKzd45xB1NFzxxgnEEp6k - as you can see, there's nothing that stands out particularly as worse than usual.

**Reply by davidv1992** [reply]
Hmm, I don't have a good explanation yet, and just realized there are two things we don't currently output as metrics for time sources, namely how the upstream reports its uncertainty and delay (which are taken into account in the combined uncertainty of the system). I'm writing a PR to make these available, but unless you see something unusual in log files or monitoring of those upstream servers I think there is no good route for diagnosing the uncertainty bump here.

**Reply by davidv1992** [reply]
Although looking in a bit more detail at the new data, multiple of the time source ip's occur multiple times, suggesting the time sources aren't entirely stable.

**Reply by paulgear** [reply]
> Although looking in a bit more detail at the new data, multiple of the time source ip's [sic] occur multiple times, suggesting the time sources aren't entirely stable.

That's just an artefact of how grafana represents them before and after the upgrade to 1.0.0-rc3.  If you zoom the graph to only include data before 2023-09-20 08:00 UTC or only after that, you'll see only one of each source.
